File: tools/JavaScript/t4-javascript-api/src/definitions/chartdata/ChartFormatAggr.js
/**
 * Port of `com.t4login.definitions.chartdata.ChartFormatAggr`.
 *
 * - `Bar` (plain data object): single aggregated OHLCV bar.
 * - `MarketDefinition`: implements IMarketConversion with lazy
 *   min-price-increment and VPT derivation.
 * - CTAG_* constants for the aggregated (T4BinAggr) binary format.
 */

import { Decimal, HALF_EVEN } from '../../decimal.js';
import { Price, Scale } from '../priceconversion/Price.js';
import { VPT } from '../priceconversion/VPT.js';

// --- T4BinAggr tags ---------------------------------------------------------
export const CVAL_T4BINAGGR_VERSION = 1;

export const CTAG_SOF = 1;
export const CTAG_MARKET_DEFINITION = 2;
export const CTAG_MARKET_SWITCH = 3;
export const CTAG_TRADEDATE_SWITCH = 4;
export const CTAG_BAR_DELTA = 10;
export const CTAG_BAR = 11;
export const CTAG_MARKET_MODE = 20;
export const CTAG_OPEN_INTEREST = 21;
export const CTAG_SETTLEMENT_PRICE = 22;

// --- Bar --------------------------------------------------------------------

/**
 * Aggregated OHLCV bar.
 *
 * Field names preserve PascalCase from Java/Python for 1:1 parity.
 */
export class Bar {
  constructor({
    TradeDate,
    Time,
    CloseTime,
    MarketID,
    OpenPrice,
    HighPrice,
    LowPrice,
    ClosePrice,
    Volume,
    VolumeAtBid,
    VolumeAtOffer,
    Trades,
    TradesAtBid,
    TradesAtOffer,
  }) {
    this.TradeDate = TradeDate;
    this.Time = Time;
    this.CloseTime = CloseTime;
    this.MarketID = MarketID;
    this.OpenPrice = OpenPrice;
    this.HighPrice = HighPrice;
    this.LowPrice = LowPrice;
    this.ClosePrice = ClosePrice;
    this.Volume = Volume;
    this.VolumeAtBid = VolumeAtBid;
    this.VolumeAtOffer = VolumeAtOffer;
    this.Trades = Trades;
    this.TradesAtBid = TradesAtBid;
    this.TradesAtOffer = TradesAtOffer;
  }
}

// --- MarketDefinition -------------------------------------------------------

const _QUANTUM = new Decimal(10).pow(-Scale);

/**
 * Market parameters for price conversion (mirrors Java inner class).
 * Implements the IMarketConversion contract via explicit getters.
 */
export class MarketDefinition {
  constructor({
    MarketID,
    Numerator,
    Denominator,
    PriceCode,
    TickValue,
    VPT_str = '',
    MinCabPrice = null,
  }) {
    this.MarketID = MarketID;
    this.Numerator = Numerator;
    this.Denominator = Denominator;
    this.PriceCode = PriceCode;
    this.TickValue = TickValue instanceof Decimal ? TickValue : new Decimal(TickValue);
    this.VPT_str = VPT_str;
    this.MinCabPrice = MinCabPrice;

    const incr = new Decimal(this.Numerator)
      .div(new Decimal(this.Denominator))
      .toDecimalPlaces(Scale, HALF_EVEN);
    this._minPriceIncrement = new Price(incr);

    if ((this.VPT_str && this.VPT_str.length > 0) || this.MinCabPrice != null) {
      this._vpt = new VPT(this.VPT_str, this.MarketID, this._minPriceIncrement, this.MinCabPrice);
    } else {
      this._vpt = null;
    }
  }

  // --- IMarketConversion -------------------------------------------------
  getMarketId() { return this.MarketID; }
  getDenominator() { return this.Denominator; }
  getPriceCode() { return this.PriceCode; }
  getMinPriceIncrement() { return this._minPriceIncrement; }
  getVpt() { return this._vpt; }
  getMinCabPrice() { return this.MinCabPrice; }
  getRealDecimals() { return 0; }
  getClearingDecimals() { return 0; }
  getPointValue() { return new Decimal(0); }
  getYieldYears() { return 0; }
  getYieldParValue() { return 0; }
  getYieldRate() { return 0; }
  getYieldValueDenominator() { return 0; }
  getYieldRedemption() { return 0; }
  getYieldPaymentsPerYear() { return 0; }
  getYieldBasis() { return 0; }
}

File: tools/Rust/t4-Rust-api/src/chart_format_aggr.rs
//! T4BinAggr tags, the [`Bar`] data object, and [`MarketDefinition`].
//!
//! Port of `chart_format_aggr.hpp`.

use crate::big_int::BigInt;
use crate::decimal::Decimal;
use crate::market::MarketConversion;
use crate::n_date_time::NDateTime;
use crate::price::Price;
use crate::vpt::Vpt;

/// T4BinAggr record tag constants.
pub mod tags {
    pub const T4BINAGGR_VERSION: i32 = 1;
    pub const SOF: i32 = 1;
    pub const MARKET_DEFINITION: i32 = 2;
    pub const MARKET_SWITCH: i32 = 3;
    pub const TRADEDATE_SWITCH: i32 = 4;
    pub const BAR_DELTA: i32 = 10;
    pub const BAR: i32 = 11;
    pub const MARKET_MODE: i32 = 20;
    pub const OPEN_INTEREST: i32 = 21;
    pub const SETTLEMENT_PRICE: i32 = 22;
}

/// A single aggregated OHLCV bar. Field names match the reference sources.
#[derive(Clone, Debug, Default)]
pub struct Bar {
    pub trade_date: NDateTime,
    pub time: NDateTime,
    pub close_time: NDateTime,
    pub market_id: String,
    pub open_price: Price,
    pub high_price: Price,
    pub low_price: Price,
    pub close_price: Price,
    pub volume: i32,
    pub volume_at_bid: i32,
    pub volume_at_offer: i32,
    pub trades: i32,
    pub trades_at_bid: i32,
    pub trades_at_offer: i32,
}

/// Market parameters for price conversion (mirrors the Java inner class).
#[derive(Clone, Debug, Default)]
pub struct MarketDefinition {
    pub market_id: String,
    pub numerator: i32,
    pub denominator: i32,
    pub price_code: String,
    pub tick_value: Decimal,
    pub vpt_str: String,
    pub min_cab_price: Option<Price>,
    min_price_increment: Price,
    vpt: Option<Vpt>,
}

impl MarketDefinition {
    /// Build a market definition, eagerly computing the min price increment and
    /// (if the market publishes one) its VPT.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        market_id: String,
        numerator: i32,
        denominator: i32,
        price_code: String,
        tick_value: Decimal,
        vpt_str: String,
        min_cab_price: Option<Price>,
    ) -> Self {
        let min_price_increment = Price::new(Decimal::divide_int(
            &BigInt::from_i64(numerator as i64),
            denominator as u64,
            Price::SCALE,
        ));
        let vpt = if !vpt_str.is_empty() || min_cab_price.is_some() {
            Some(Vpt::new(vpt_str.clone()))
        } else {
            None
        };
        MarketDefinition {
            market_id,
            numerator,
            denominator,
            price_code,
            tick_value,
            vpt_str,
            min_cab_price,
            min_price_increment,
            vpt,
        }
    }
}

impl MarketConversion for MarketDefinition {
    fn denominator(&self) -> i64 {
        self.denominator as i64
    }

    fn min_price_increment(&self) -> Price {
        self.min_price_increment.clone()
    }

    fn vpt(&self) -> Option<&Vpt> {
        self.vpt.as_ref()
    }
}
